lib/fetch-request.js

```diff
@@ -338,44 +376,51 @@ export default class FetchRequest {
+   * @param {Array<string>} fields Properties of each object to retrieve
+   * @returns {Array<object>} Array of requested objects
    */
-  async getAndEmbedCollectionObjects(items, embedProp) {
-    // Retrieve the specified non-embedded objects from the server. Any objects that return an
-    // error get filtered out.
-    const paths = _.uniq(items.map((item) => item[embedProp]));
-    const objects = await this.getMultipleObjects(paths, null);
-    const successfulObjects = objects.filter((object) => object);
-
-    // Embed the retrieved objects in the given items. Use a simple cache using the embedded
-    // property path as a key to reduce the number of searches through the fetched objects.
-    const cache = {};
-    items.forEach((item) => {
-      const objectPath = item[embedProp];
-      let cachedObject = cache[objectPath];
-      if (!cachedObject) {
-        // Cache miss: add requested object to the cache with its path as key. Any objects
-        // containing fetch errors won't have an @id, so they don't get cached nor embedded.
-        const matchingObject = successfulObjects.find(
-          (object) => object["@id"] === objectPath
-        );
-        if (matchingObject) {
-          cache[objectPath] = matchingObject;
-          cachedObject = cache[objectPath];
-        }
-      }
+  async getMultipleObjectsBulk(paths, defaultErrorValue, fields) {
+    logRequest("getMultipleObjectsBulk", `[${paths.join(", ")}]`);
 
-      // Set the embedded property to the (possibly newly) cached object, or null if we couldn't
-      // retrieve the corresponding object.
-      item[embedProp] = cachedObject || null;
-    });
+    if (paths.length > 0) {
+      // Generate the query string for the needed fields of each object.
+      const fieldQuery = fields.map((field) => `field=${field}`).join("&");
+
+      // Break the paths into groups of MAX_PATH_GROUP_SIZE, each group mapping to a data-provider
+      // request. This reduces the lengths of the query strings to fit within the data provider's
+      // limits.
+      const pathGroups = this.#pathsIntoPathGroups(paths, fieldQuery.length);
+
+      // For each group of paths, request the objects as search results. Send these requests in
+      // parallel.
+      const results = await Promise.all(
+        pathGroups.map(async (group) => {
+          const pathQuery = group.map((path) => `@id=${path}`).join("&");
+          const query = `${fieldQuery}&${pathQuery}`;
+          const response = await this.getObject(
+            `/search/?${query}&limit=${group.length}`,
+            null
+          );
+          if (response?.["@graph"]) {
+            // Add the results to the array of results.
+            return response["@graph"];
+          }
+
+          // The response was an error. Add an empty array to the path groups, so that segment of
+          // results gets skipped.
+          return [];
```

Comment:
I wonder if this is what we want to do in case of an error. It looks like we just swallow the error if there is any?

I see a few different ways to deal with errors here:
1. We could collect valid results along with errors. I'll pseudo-code the types like rust, it's how I often think about types haha (Vec is vector, like list or a growable array). So on line 401, `results` would have type: `Vec<Result<Vec<T>, Error>`, and then the consumer would decide how to sort through them. But maybe if we had enough info in the error returned (like what `@id`s was being fetched when some error occurred, we could kinda handle it ourselves, maybe like:
2. `getMultipleObjectsBulk` would have return type: `(Vec<T>, Vec<E>)` collecting all the valid results and all errors into a tuple. Then callers could see what (if any) calls erred and what the error was. But they'd still just have the bulk valid results if the caller wanted to ignore the error.
3. It seems like we don't want this, but I guess we could bail on the entire operation, and have a return type: `Result<Vec<T>, E>` where E is the first error to occur?

Reply:
In a discussion, we decided we do actually want to do 3, so I'll say "request changes" to reflect that.